### ISSUE TITLE
ImportError ~ psycopg2 when running deploy task on Heroku

### ISSUE THREAD

**Issue by fkromer** [miguelgrinberg/flasky#206]
When I run the deploy task of my code (~ tag 17c) on Heroku following your book instructions [book, 1st edition, p. 219-221/226] I get the following error:

    ...
    (venv)florian@florian-desktop ~/ws_github/npkromer $ heroku run python manage.py deploy
    Running python manage.py deploy on ⬢ npkromer... up, run.7867 (Free)
    Traceback (most recent call last):
      File "manage.py", line 70, in <module>
        manager.run()
      File "/app/.heroku/python/lib/python2.7/site-packages/flask_script/__init__.py", line 405, in run
        result = self.handle(sys.argv[0], sys.argv[1:])
      File "/app/.heroku/python/lib/python2.7/site-packages/flask_script/__init__.py", line 384, in handle
        return handle(app, *positional_args, **kwargs)
      File "/app/.heroku/python/lib/python2.7/site-packages/flask_script/commands.py", line 145, in handle
        return self.run(*args, **kwargs)
      File "manage.py", line 63, in deploy
        upgrade()
      File "/app/.heroku/python/lib/python2.7/site-packages/flask_migrate/__init__.py", line 98, in upgrade
        command.upgrade(config, revision, sql = sql, tag = tag)
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/command.py", line 124, in upgrade
        script.run_env()
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/script.py", line 199, in run_env
        util.load_python_file(self.dir, 'env.py')
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/util.py", line 199, in load_python_file
        module = load_module(module_id, path)
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/compat.py", line 55, in load_module
        mod = imp.load_source(module_id, path, fp)
      File "migrations/env.py", line 72, in <module>
        run_migrations_online()
      File "migrations/env.py", line 55, in run_migrations_online
        poolclass=pool.NullPool)
      File "/app/.heroku/python/lib/python2.7/site-packages/sqlalchemy/engine/__init__.py", line 408, in engine_from_config
        return create_engine(url, **options)
      File "/app/.heroku/python/lib/python2.7/site-packages/sqlalchemy/engine/__init__.py", line 386, in create_engine
        return strategy.create(*args, **kwargs)
      File "/app/.heroku/python/lib/python2.7/site-packages/sqlalchemy/engine/strategies.py", line 74, in create
        dbapi = dialect_cls.dbapi(**dbapi_args)
      File "/app/.heroku/python/lib/python2.7/site-packages/sqlalchemy/dialects/postgresql/psycopg2.py", line 541, in dbapi
        import psycopg2
    ImportError: No module named psycopg2

I use Python3.4.3 locally, not Python2.7...

BTW: The command on p. 220 `heroku addons: heroku-postgresql:dev` is deprecated. Use `heroku addons:create heroku-postgresql:hobby-dev` instead. There is no need for `heroku pg:promote HEROKU_PPOSTGRESQL_BROWN_URL` anymore...

**Comment by miguelgrinberg** [comment]
Did you restore psycopg2 in the requirements file? That is required to talk to the Postgres database.

**Comment by fkromer** [comment]
Yes, I did restore psycopg2 in the requirement files.

**Comment by miguelgrinberg** [comment]
@fkromer the error that you are getting suggests that psycopg2 isn't getting installed though. You need to review the `git push` that deployed to heroku, the output of `pip install -r requirements.txt` is going to be there, and you should see what was installed.

**Comment by fkromer** [comment]
psycopg2 seems to be installed:

    (venv)florian@florian-desktop ~/ws_github/npkromer $ git push heroku master
    Counting objects: 9, done.
    Delta compression using up to 2 threads.
    Compressing objects: 100% (5/5), done.
    Writing objects: 100% (5/5), 552 bytes | 0 bytes/s, done.
    Total 5 (delta 2), reused 0 (delta 0)
    remote: Compressing source files... done.
    remote: Building source:
    remote: 
    remote: -----> Python app detected
    remote:      $ pip install -r requirements.txt
    remote:        Collecting psycopg2==2.5.1 (from -r requirements.txt (line 4))
    remote:          Downloading psycopg2-2.5.1.tar.gz (684kB)
    remote:        Installing collected packages: psycopg2
    remote:          Running setup.py install for psycopg2: started
    remote:            Running setup.py install for psycopg2: finished with status 'done'
    remote:        Successfully installed psycopg2-2.5.1
    remote: 
    remote: -----> Discovering process types
    remote:        Procfile declares types -> web
    remote: 
    remote: -----> Compressing...
    remote:        Done: 40.5M
    remote: -----> Launching...
    remote:        Released v9
    remote:        https://npkromer.herokuapp.com/ deployed to Heroku
    remote: 
    remote: Verifying deploy... done.
    To https://git.heroku.com/npkromer.git
       f9f9b08..0959698  master -> master
    (venv)florian@florian-desktop ~/ws_github/npkromer $

**Comment by fkromer** [comment]
but there seems to be some problem with the db files:

    (venv)florian@florian-desktop ~/ws_github/npkromer $ heroku run python manage.py deploy
    Running python manage.py deploy on ⬢ npkromer... up, run.7017 (Free)
    INFO  [alembic.migration] Context impl PostgresqlImpl.
    INFO  [alembic.migration] Will assume transactional DDL.
    Traceback (most recent call last):
      File "manage.py", line 70, in <module>
        manager.run()
      File "/app/.heroku/python/lib/python2.7/site-packages/flask_script/__init__.py", line 405, in run
        result = self.handle(sys.argv[0], sys.argv[1:])
      File "/app/.heroku/python/lib/python2.7/site-packages/flask_script/__init__.py", line 384, in handle
        return handle(app, *positional_args, **kwargs)
      File "/app/.heroku/python/lib/python2.7/site-packages/flask_script/commands.py", line 145, in handle
        return self.run(*args, **kwargs)
      File "manage.py", line 63, in deploy
        upgrade()
      File "/app/.heroku/python/lib/python2.7/site-packages/flask_migrate/__init__.py", line 98, in upgrade
        command.upgrade(config, revision, sql = sql, tag = tag)
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/command.py", line 124, in upgrade
        script.run_env()
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/script.py", line 199, in run_env
        util.load_python_file(self.dir, 'env.py')
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/util.py", line 199, in load_python_file
        module = load_module(module_id, path)
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/compat.py", line 55, in load_module
        mod = imp.load_source(module_id, path, fp)
      File "migrations/env.py", line 72, in <module>
        run_migrations_online()
      File "migrations/env.py", line 65, in run_migrations_online
        context.run_migrations()
      File "<string>", line 7, in run_migrations
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/environment.py", line 652, in run_migrations
        self.get_context().run_migrations(**kw)
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/migration.py", line 210, in run_migrations
        self):
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/command.py", line 113, in upgrade
        return script._upgrade_revs(revision, rev)
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/script.py", line 179, in _upgrade_revs
        for script in reversed(list(revs))
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/script.py", line 161, in _iterate_revisions
        lower = self.get_revision(lower)
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/script.py", line 98, in get_revision
        return self._revision_map[id_]
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/util.py", line 233, in __get__
        obj.__dict__[self.__name__] = result = self.fget(obj)
      File "/app/.heroku/python/lib/python2.7/site-packages/alembic/script.py", line 208, in _revision_map
        for file_ in os.listdir(self.versions):
    OSError: [Errno 2] No such file or directory: 'migrations/versions'

**Comment by miguelgrinberg** [comment]
Okay, this new error makes more sense. You have to include the `migrations` folder with all of its contents in the git repository that you push to heroku. This is where the migration scripts are stored, those need to be on heroku for the database to be created and upgraded.

**Comment by fkromer** [comment]
I did get it running on Heroku. Actually there were 2 issues:

1) The first issue ~ `ImportError: No module named psycopg2` occured when trying to install my version of the **flasky tag 17c** like described in the [book, 1st edition, p. 225] with `pip install -r requiremetns.txt`. Actually I had to remove the install instruction `psycopg2==2.5.1` in the [overall requirement file of tag 17c](https://github.com/miguelgrinberg/flasky/blob/17c/requirements.txt) to beeing able to install everything properly for local testing with heroku local. Is this statement required in this file as the dependency is included in the [heroku speficic requirement file of tag 17c](https://github.com/miguelgrinberg/flasky/blob/17c/requirements/heroku.txt)?

2) The second database related issue was because I did not merge the tag 17c and mine *.gitignore* file... I had some stupid error in my *.gitignore* file.
